ssz_serialization/merkleization.nim

```diff
@@ -746,3 +1169,217 @@ func hash_tree_root*(x: auto): Digest =
+  result = toSeq(indices.low .. indices.high)
+  let idx = cast[ptr UncheckedArray[GeneralizedIndex]](unsafeAddr indices[0])
+  result.sort do (x, y: int) -> int:
+    cmpDepthFirst(idx[x], idx[y])
+
+func merkleizationLoopOrder(
+    indices: openArray[GeneralizedIndex]): seq[int] =
+  when nimvm:
+    merkleizationLoopOrderNimvm(indices)
+  else:
+    merkleizationLoopOrderRegular(indices)
+
+func validateIndices(
+    indices: openArray[GeneralizedIndex],
+    loopOrder: seq[int]): Result[void, string] =
+  var
+    prev = indices[loopOrder[0]]
+    prevLayer = log2trunc(prev)
+  if prev < 1.GeneralizedIndex: return err("Invalid generalized index.")
+  for i in 1 .. loopOrder.high:
+    let
+      curr = indices[loopOrder[i]]
+      currLayer = log2trunc(curr)
+      indicesOverlap =
+        if currLayer < prevLayer:
+          (prev shr (prevLayer - currLayer)) == curr
+        elif currLayer > prevLayer:
+          (curr shr (currLayer - prevLayer)) == prev
+        else:
+          curr == prev
+    if indicesOverlap:
+      return err("Given indices cover some leafs multiple times.")
+    prev = curr
+    prevLayer = currLayer
+  ok()
+
+func hash_tree_root*(
+    x: auto,
+    indices: openArray[GeneralizedIndex],
+    roots: var openArray[Digest]): Result[void, string] =
+  doAssert indices.len == roots.len
+  if indices.len == 0:
+    ok()
+  elif indices.len == 1 and indices[0] == 1.GeneralizedIndex:
+    roots[0] = hash_tree_root(x)
+    ok()
+  else:
+    let loopOrder = merkleizationLoopOrder(indices)
+    ? validateIndices(indices, loopOrder)
+    let slice = 0 ..< loopOrder.len
+    hash_tree_root_multi(x, indices, roots, loopOrder, slice)
+
+func hash_tree_root*(
+    x: auto,
+    indices: static openArray[GeneralizedIndex],
+    roots: var openArray[Digest]): Result[void, string] =
+  doAssert indices.len == roots.len
+  when indices.len == 0:
+    ok()
+  else:
+    when indices.len == 1 and indices[0] == 1.GeneralizedIndex:
+      roots[0] = hash_tree_root(x)
+      ok()
+    else:
+      const
+        loopOrder = merkleizationLoopOrder(indices)
+        v = validateIndices(indices, loopOrder)
+      when v.isErr:
+        err(v.error)
+      else:
+        const slice = 0 ..< loopOrder.len
+        hash_tree_root_multi(x, indices, roots, loopOrder, slice)
+
+func hash_tree_root*(
+    x: auto,
+    indices: openArray[GeneralizedIndex]
+): Result[seq[Digest], string] =
+  if indices.len == 0:
+    ok(newSeq[Digest](0))
+  elif indices.len == 1 and indices[0] == 1.GeneralizedIndex:
+    ok(@[hash_tree_root(x)])
+  else:
+    let loopOrder = merkleizationLoopOrder(indices)
+    ? validateIndices(indices, loopOrder)
+    let slice = 0 ..< loopOrder.len
+    var roots = newSeq[Digest](indices.len)
+    ? hash_tree_root_multi(x, indices, roots, loopOrder, slice)
+    ok(roots)
+
+func hash_tree_root*(
+    x: auto,
+    indices: static openArray[GeneralizedIndex]
+): auto =
+  when indices.len == 0:
```

Comment:
The code may be more clear if you place `type ResultType = Result[array[indices.len, Digest], string]` on the first line and then refer to it in the rest of the body.

Reply:
Oh, that's very clean indeed! Have switched to using this.